### PR TITLE
[SPARK-35282][SQL] Support AQE side shuffled hash join formula using rule

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

The main code change is:
* Change rule `DemoteBroadcastHashJoin` to `DynamicJoinSelection` and add shuffle hash join selection code.
* Specify a join strategy hint `PREFER_SHUFFLE_HASH` or `SHUFFLE_HASH` if AQE think a join can be converted to SHJ.
* Skip `preferSortMerge` config check in AQE side if a join can be converted to SHJ.

### Why are the changes needed?

Use AQE runtime statistics to decide if we can use shuffled hash join instead of sort merge join. Currently, the formula of shuffled hash join selection dose not work due to the dymanic shuffle partition number.

Add a new config `spark.sql.adaptive.maxShuffledHashJoinLocalMapThreshold` to decide if join can be converted to shuffled hash join safely.

### Does this PR introduce _any_ user-facing change?

Yes, add a new config.

### How was this patch tested?

Add test.

### CODE REVIEW

**Comment by ulysses-you** [comment]
An another approach of [SPARK-35282](https://issues.apache.org/jira/browse/SPARK-35282) using rule.  Do you have time to take a look ?  cc @maropu @cloud-fan @c21 @dongjoon-hyun @advancedxy

**Comment by HyukjinKwon** [comment]
cc @maryannxue

**Comment by ulysses-you** [comment]
The reason of failed tests is the Join strategy from `SMJ` change to `SHJ` since we skip the config `preferSortMerge`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -598,6 +598,16 @@ object SQLConf {
       .bytesConf(ByteUnit.BYTE)
       .createOptional
 
+  val ADAPTIVE_SHUFFLE_HASH_JOIN_LOCAL_MAP_THRESHOLD =
+    buildConf("spark.sql.adaptive.shuffledHashJoinLocalMapThreshold")
+      .doc("Configures the maximum size in bytes for per partition that can be allowed to build " +
+        "local hash map. If all the partition size not larger than this threshold, join " +
+        "selection may use shuffled hash join instead of sort merge join when " +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:605]: `when` -> `even if`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -598,6 +598,16 @@ object SQLConf {
...
+        s"${PREFER_SORTMERGEJOIN.key} is false.")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:606]: really? Shall we use shuffle hash join regardless of this `PREFER_SORTMERGEJOIN` config?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  private def selectJoinStrategy(plan: LogicalPlan): Option[JoinStrategyHint] = plan match {
+    case LogicalQueryStage(_, stage: ShuffleQueryStageExec) if stage.resultOption.get().isDefined
+      && stage.mapStats.isDefined =>
+      var strategies = Seq[JoinStrategyHint]()
+      if (shouldDemoteBroadcastHashJoin(stage)) {
+        strategies = strategies ++ Seq(NO_BROADCAST_HASH)
+      }
+      if (preferShuffledHashJoin(stage)) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:52]: are these two join hints exclusive? shall we use `else if` here?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:52]: I think `PREFER_SHUFFLE_HASH` means we shouldn't do broadcast join.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:606]: yeah, right. I missed update this docs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala (hunk 3 above)
Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:52]: Not exclusive. My thought is `PREFER_SHUFFLE_HASH` is not a determining hint which is same as `NO_BROADCAST_HASH`. `NO_BROADCAST_HASH` is for `BHJ` and other shuffle based join, and `PREFER_SHUFFLE_HASH` is for `SMJ` and `SHJ`.

So one join can be a `BHJ`  even if we have a `PREFER_SHUFFLE_HASH` hint but never be a `SMJ`. In this idea, we can inject some join recommendation for `JoinSelection`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala [hunk 4]
```diff
@@ -172,6 +190,14 @@ case object NO_BROADCAST_HASH extends JoinStrategyHint {
   override def hintAliases: Set[String] = Set.empty
 }
 
+/**
+ * An internal hint to encourage shuffle hash join, used by adaptive query execution.
+ */
+case object PREFER_SHUFFLE_HASH extends JoinStrategyHint {
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala:196]: Sorry if I miss anything, why we cannot reuse the existing hint `SHUFFLE_HASH`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -598,6 +598,16 @@ object SQLConf {
       .bytesConf(ByteUnit.BYTE)
       .createOptional
 
+  val ADAPTIVE_SHUFFLE_HASH_JOIN_LOCAL_MAP_THRESHOLD =
+    buildConf("spark.sql.adaptive.shuffledHashJoinLocalMapThreshold")
+      .doc("Configures the maximum size in bytes for per partition that can be allowed to build " +
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:603]: nit: unnecessary `for`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -598,6 +598,16 @@ object SQLConf {
...
+        "local hash map. If all the partition size not larger than this threshold, join " +
+        "selection may use shuffled hash join instead of sort merge join regardless of the " +
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:605]: I feel `may` is quite vague here. Can't we just say `prefer to use shuffled hash join instead of sort merge join regardless of ...`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala [hunk 7]
```diff
@@ -255,12 +255,28 @@ trait JoinSelectionHelper {
     val buildLeft = if (hintOnly) {
       hintToShuffleHashJoinLeft(hint)
     } else {
-      canBuildLocalHashMapBySize(left, conf) && muchSmaller(left, right)
+      if (hintToPreferShuffleHashJoinLeft(hint)) {
```
Comment (c21) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:258]: what about the case when both left and right sides meet requirement of AQE shuffled hash join (i.e. each partition size < threshold) ? Here we always build left. Shall we build right in case right side is smaller than left side?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEOptimizer.scala [hunk 8]
```diff
@@ -28,7 +28,7 @@ import org.apache.spark.util.Utils
 class AQEOptimizer(conf: SQLConf) extends RuleExecutor[LogicalPlan] {
   private val defaultBatches = Seq(
     Batch("Eliminate Unnecessary Join", Once, EliminateUnnecessaryJoin),
-    Batch("Demote BroadcastHashJoin", Once, DemoteBroadcastHashJoin)
+    Batch("Dynamic JoinSelection", Once, DynamicJoinSelection)
```
Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEOptimizer.scala:31]: nit: `Dynamic Join Selection`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala (hunk 4 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala:196]: If we want to make a `SHUFFLE_HASH` hint,  we need to do more check.

The priority of join strategy is in `JoinSelection` which is `BHJ` > `SMJ` > `SHJ`. So `SHUFFLE_HASH` need a check to make sure we cannot build a `BHJ`. The the idea of `PREFER_SHUFFLE_HASH` is for skipping this extra check at AQE side and let `JoinSelection` to decide the final join strategy.

Logiacally, we can mix `JoinSelection` in AQE and use redundant code to choose join strategy directly using a specific hint. But the issue is the code maintain @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala (hunk 7 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:258]: Yeah,  we have already chose a smaller one since the `LogicalQueryStage` has override the `computeStats`. And see the code in `joins.getBuildSide`:
```
private def getBuildSide(
      canBuildLeft: Boolean,
      canBuildRight: Boolean,
      left: LogicalPlan,
      right: LogicalPlan): Option[BuildSide] = {
    if (canBuildLeft && canBuildRight) {
      // returns the smaller side base on its estimated physical size, if we want to build the
      // both sides.
      Some(getSmallerSide(left, right))
    } else if (canBuildLeft) {
      Some(BuildLeft)
    } else if (canBuildRight) {
      Some(BuildRight)
    } else {
      None
    }
  }
```

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:258]: Add an build side check if you worry about this.  [c747a23](https://github.com/apache/spark/pull/32550/commits/c747a23be0e59bcb7edb6e9bb377d2aed7248c96)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:52]: I think the major difference between this new `PREFER_SHUFFLE_HASH` and `SHUFFLE_HASH` hints is: `PREFER_SHUFFLE_HASH` forces to do shuffle hash join even if we can broadcast.

I'd really like to avoid `JoinStrategyHintCollection`, which adds complexity to the join hint framework.

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala]: @cloud-fan How about this updating ? Has simplified code about hints.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 9]
```diff
@@ -598,6 +598,16 @@ object SQLConf {
       .bytesConf(ByteUnit.BYTE)
       .createOptional
 
+  val ADAPTIVE_SHUFFLE_HASH_JOIN_LOCAL_MAP_THRESHOLD =
+    buildConf("spark.sql.adaptive.shuffledHashJoinLocalMapThreshold")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:602]: `shuffledHashJoinLocalMapSizeThreshold`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+ * This optimization rule includes two join selection:
+ *   1. detects a join child that can not broadcast and every partition size less than local map
+ *      threshold. Adds a shuffled_hash hint to force being shuffle hash join.
+ *   2. detects a join child that has a high ratio of empty partitions.
+ *      Adds a no_broadcast_hash_join hint to avoid it being broadcast.
+ */
+object DynamicJoinSelection extends Rule[LogicalPlan] with JoinSelectionHelper {
+
+  private def shouldDemoteBroadcastHashJoin(stage: ShuffleQueryStageExec): Boolean = {
+    val mapStats = stage.mapStats.get
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:34]: nit: both this function and the next function should take `mapStats` as the input, and let the caller side do `stage.mapStats.get`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    val partitionCnt = mapStats.bytesByPartitionId.length
+    val nonZeroCnt = mapStats.bytesByPartitionId.count(_ > 0)
+    partitionCnt > 0 && nonZeroCnt > 0 &&
+      (nonZeroCnt * 1.0 / partitionCnt) < conf.nonEmptyPartitionRatioForBroadcastJoin
+  }
+
+  private def preferShuffledHashJoin(plan: LogicalPlan, stage: ShuffleQueryStageExec): Boolean = {
+    val localMapThreshold = conf.shuffleHashJoinLocalMapThreshold
+    !canBroadcastBySize(plan, conf) &&
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:43]: So what we really need is a hint that prefers shuffle hash join over SMJ, but still does broadcast if applicable.

I think it's more robust to create a new hint for this purpose, then we don't need to check canBroadcast here.

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:43]: yea, this is my origin thought of `PREFER_SHUFFLE_JOIN` that try to avoid touch broadcast join at this place. So follow this idea, it is not exclusive with `NO_BROADCAST_HASH` then I used a `JoinStrategyHintCollection` to keep these two hints.

Do you think which version is better ?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:43]: OK I got your point now. For a certain join, we may want to demote the broadcast join, and we may also want to switch to shuffle hash join. These two can happen at the same time.

However, if we demote the broadcast join, many partitions are empty and I don't think the join algorithm matters anymore. We can just say:
1. If  `NO_BROADCAST_HASH`, we should use SMJ
2. If `PREFER_SHUFFLE_HASH`, we should use shuffle hash join if broadcast join is not applicable.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:43]: or we can have 3 hints for different cases
1. if we want to demote and want to use shuffle hash join, then use `SHUFFLE_HASH` hint
2. if we don't want to demote and want to use shuffle hash join, then use `PREFER_SHUFFLE_HASH` hint
3. if we want to demote and don't want to use shuffle hash join, then use `NO_BROADCAST_HASH` hint

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:43]: the last idea +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala [hunk 12]
```diff
@@ -253,12 +253,28 @@ trait JoinSelectionHelper {
...
+      }
     }
     val buildRight = if (hintOnly) {
       hintToShuffleHashJoinRight(hint)
     } else {
-      canBuildLocalHashMapBySize(right, conf) && muchSmaller(right, left)
+      if (hintToPreferShuffleHashJoinRight(hint)) {
+        true
+      } else {
+        if (!conf.preferSortMergeJoin) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala:272]: is this change related to this PR? the old code was `canBuildLocalHashMapBySize(right, conf) && muchSmaller(right, left)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+          newHint = newHint.copy(leftHint =
+            Some(hint.leftHint.getOrElse(HintInfo()).copy(strategy = Some(strategy))))
+        }
+      }
+      if (!hint.rightHint.exists(_.strategy.isDefined)) {
+        selectJoinStrategy(right).foreach { strategy =>
+          newHint = newHint.copy(rightHint =
+            Some(hint.rightHint.getOrElse(HintInfo()).copy(strategy = Some(strategy))))
+        }
+      }
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:79]: It seems possible that both sides have `PREFER_SHUFFLE_HASH` hint. How do we decide the build side?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:79]: The same question applies to `SHUFFLE_HASH` as well, but that's an existing problem.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:79]: e.g. `SHUFFLE_HASH(t1, t2)` with an inner join, do we always pick the left side as the build side, even if the right side is smaller?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:79]: Ah sorry I misread the code. At the end we call `getBuildSide`, and pick the smaller side as the build side, if both sides are be the build side.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.MapOutputStatistics
+import org.apache.spark.sql.catalyst.plans.logical.{HintInfo, Join, JoinStrategyHint, LogicalPlan, NO_BROADCAST_HASH, PREFER_SHUFFLE_HASH, SHUFFLE_HASH}
+import org.apache.spark.sql.catalyst.rules.Rule
+
+/**
+ * This optimization rule includes two join selection:
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:25]: two -> three

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+ *   1. detects a join child that has a high ratio of empty partitions and adds a
+ *      NO_BROADCAST_HASH hint to avoid it being broadcast.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:27]: let's add the reason: `..., as shuffle join is faster in this case: many tasks complete immediately since one join side is empty.`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+ *   2. detects a join child that every partition size less than local map threshold and adds a
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:28]: `every partition size is less than`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 17]
```diff
@@ -1707,4 +1707,50 @@ class AdaptiveQueryExecSuite
...
+      spark.sparkContext.parallelize(
+        (1 to 10).map(i => TestData(i, i.toString)), 5)
+        .toDF("c1", "c2").createOrReplaceTempView("t2")
+
+      // t1 partition size: [926, 729, 731]
+      // t2 partition size: [318, 120, 0]
+      withSQLConf(SQLConf.SHUFFLE_PARTITIONS.key -> "3",
+        SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1",
+        SQLConf.PREFER_SORTMERGEJOIN.key -> "true") {
+        withSQLConf(SQLConf.ADAPTIVE_SHUFFLE_HASH_JOIN_LOCAL_MAP_THRESHOLD.key -> "318") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1744]: It's a bit fragile to match the size exactly. How about `300`, `400` and `1000`?

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala (hunk 13 above)
Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:80]: yea, and has already added a test before [c747a23](https://github.com/apache/spark/pull/32550/commits/c747a23be0e59bcb7edb6e9bb377d2aed7248c96)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala (hunk 17 above)
Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala]: updated

**Comment by ulysses-you** [comment]
@cloud-fan  I just found current implemtation has some potential issue.

After this PR, we can use `shuffledHashJoinLocalMapThreshold` to deicde if a join can be SHJ. However after AQE Optimizer, `CoalesceShufflePartitions` would coalesce partition size if it's possible. Then the check of `shuffledHashJoinLocalMapThreshold` would be broken. The result per-partition size would be `advisoryPartitionSizeInBytes` which can be bigger than `shuffledHashJoinLocalMapThreshold`.

yea, maybe we can say if we want to enable this feature then should tune the both `shuffledHashJoinLocalMapThreshold` and `advisoryPartitionSizeInBytes` config to the same value. But at least we should clarify that in docs ?

Or, if we want to tune that thing automatically, we need to add a check in `CoalesceShufflePartitions` and use the `Math.min(shuffledHashJoinLocalMapThreshold, advisoryPartitionSizeInBytes)` to coalesce partition size with `ShuffledHashJoinExec` case.

**Comment by cloud-fan** [comment]
That's a good point. I think this is a general issue: if the planner creates shuffle hash join directly, we may also coalesce partitions and make the local hash map bigger.

I don't really have a good idea here. If we do have many small partitions, coalesce partitions should be beneficial even for shuffle hash join. One idea is to use the existing config `advisoryPartitionSizeInBytes` in the new rule, then we don't need to worry about the partition size after coalescing becomes much larger than we expect.

**Comment by ulysses-you** [comment]
Use `advisoryPartitionSizeInBytes` directly sounds better. How about make `shuffledHashJoinLocalMapThreshold` as  `maxShuffledHashJoinLocalMapThreshold`.

Then the new formula looks like:
```
if (advisoryPartitionSizeInBytes <= maxShuffledHashJoinLocalMapThreshold ) {
  mapStats.bytesByPartitionId.forall(_ <= advisoryPartitionSizeInBytes)
} else {
  false
}
```

**Comment by cloud-fan** [comment]
yea that's better to allow `maxShuffledHashJoinLocalMapThreshold` to be larger than `advisoryPartitionSizeInBytes`, which is more flexiable.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 18]
```diff
@@ -598,6 +598,17 @@ object SQLConf {
       .bytesConf(ByteUnit.BYTE)
       .createOptional
 
+  val ADAPTIVE_MAX_SHUFFLE_HASH_JOIN_LOCAL_MAP_THRESHOLD =
+    buildConf("spark.sql.adaptive.maxShuffledHashJoinLocalMapThreshold")
+      .doc("Configures the maximum size in bytes per partition that can be allowed to build " +
+        s"local hash map. If this value is bigger than ${ADVISORY_PARTITION_SIZE_IN_BYTES.key} " +
+        s"and all the partition size not larger than ${ADVISORY_PARTITION_SIZE_IN_BYTES.key}, " +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:605]: `not larger than this config`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 19]
```diff
@@ -598,6 +598,17 @@ object SQLConf {
...
+        "join selection prefer to use shuffled hash join instead of sort merge join regardless " +
+        s"of the value of ${PREFER_SORTMERGEJOIN.key}.")
+      .version("3.2.0")
+      .bytesConf(ByteUnit.BYTE)
+      .createOptional
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:610]: shall this fallback to `ADVISORY_PARTITION_SIZE_IN_BYTES`? `.fallbackConf(ADVISORY_PARTITION_SIZE_IN_BYTES)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+    partitionCnt > 0 && nonZeroCnt > 0 &&
+      (nonZeroCnt * 1.0 / partitionCnt) < conf.nonEmptyPartitionRatioForBroadcastJoin
+  }
+
+  private def preferShuffledHashJoin(mapStats: MapOutputStatistics): Boolean = {
+    val maxShuffledHashJoinLocalMapThreshold =
+      conf.getConf(SQLConf.ADAPTIVE_MAX_SHUFFLE_HASH_JOIN_LOCAL_MAP_THRESHOLD)
+    val advisoryPartitionSize = conf.getConf(SQLConf.ADVISORY_PARTITION_SIZE_IN_BYTES)
+    if (maxShuffledHashJoinLocalMapThreshold.exists(_ >= advisoryPartitionSize)) {
+      mapStats.bytesByPartitionId.forall(_ <= advisoryPartitionSize)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:49]: shall we use `maxShuffledHashJoinLocalMapThreshold`? Otherwise this new conf is useless.

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/DynamicJoinSelection.scala:49]: yes, it should be `maxShuffledHashJoinLocalMapThreshold` here

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 19 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:610]: If we do `fallbackConf`, many tests will fail since smj changes to shj. Not sure we should enable this feature by default.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:610]: I think we should, but let's do it in a followup PR with more discussion.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:610]: then shall we use 0 as the default value to turn it off?

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:610]: I'm ok with it

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
thanks for merging ! created followup [#32683](https://github.com/apache/spark/pull/32683)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala [hunk 21]
```diff
@@ -172,6 +172,14 @@ case object NO_BROADCAST_HASH extends JoinStrategyHint {
...
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala:178]: Can we add some user-facing docs?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 22]
```diff
@@ -598,6 +598,17 @@ object SQLConf {
       .bytesConf(ByteUnit.BYTE)
       .createOptional
 
+  val ADAPTIVE_MAX_SHUFFLE_HASH_JOIN_LOCAL_MAP_THRESHOLD =
+    buildConf("spark.sql.adaptive.maxShuffledHashJoinLocalMapThreshold")
+      .doc("Configures the maximum size in bytes per partition that can be allowed to build " +
+        "local hash map. If this value is not smaller than " +
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:604]: I would say:
- "If this value is not smaller than" => "If this value is larger than"
- "all the partition size are not larger than this config" => "all the partition size are smaller than this config"

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala (hunk 21 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/hints.scala:178]: yes, it's a internal hint and only used by AQE.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 22 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:604]: I thought about this but the formula used equal, that said it would be `If this value is larger than or equal to xxx`. Do you think it's better ?
